pkg/reconciler/apis/apiexport/apiexport_reconcile.go

```diff
@@ -203,7 +203,17 @@ func (c *controller) updateVirtualWorkspaceURLs(apiExport *apisv1alpha1.APIExpor
 		})
 	}
 
-	conditions.MarkTrue(apiExport, apisv1alpha1.APIExportVirtualWorkspaceURLsReady)
+	if len(apiExport.Status.VirtualWorkspaces) == 0 {
+		conditions.MarkFalse(
+			apiExport,
+			apisv1alpha1.APIExportVirtualWorkspaceURLsReady,
+			apisv1alpha1.NoViableShardsReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			"Unable to find any shards with virtual workspace URLs",
```

Comment:
how can that happen? If it is just a moment, not worth all this condition logic.

Reply:
I honestly have no idea. It happened to me locally. I had my server up a running for a while. I eventually created an APIExport. It had the VW URLs ready condition=true, but no URL in status. I don't know the root cause.

Reply:
cc @p0lyn0mial 

If it is not critical, would let Lukasz take a look at this next week.

Reply:
Am curious, was that old etcd state or fresh?

Reply:
I _think_ I had a preexisting set of data

Reply:
As soon as I restarted kcp (without any code changes), the APIExport got a VW URL

Reply:
I have just tried with a preexisting set and a `virtualWorkspaceURL` has been assigned.

A `virtualWorkspaceURL` is assigned during bootstrapping of the root workspace. If it hasn't been set (via a flag) it will be defaulted to the `baseURL` by an admission plugin.

Since the root `ClusterWorkspaceShard` doesn't have `annotationCreateOnlyKey` set it will always be `Updated` during bootstrapping (preexisting set). Also, any failure is eventually retired. Thoughts?

Reply:
I checked the code, nothing stands out.

Is there an easy way of searching CI for a failed test?

We could see how common the issue is with https://github.com/kcp-dev/kcp/pull/1699

Reply:
I've only seen this once, on my laptop

Reply:
would it make sense to merge https://github.com/kcp-dev/kcp/pull/1699 ?